consensus/enclave/impl/src/lib.rs

```diff
@@ -3564,4 +3603,158 @@ mod tests {
         )
         .expect("Mint txs should be valid");
     }
+
+    #[parameterized(
+        same_as_parent = {23456, 23456},
+        earlier_than_parent = {1233, 1234},
+        zero = {0, 0},
+    )]
+    fn timestamp_is_not_valid(timestamp: u64, parent_timestamp: u64) {
+        let version = 4;
+        let parent_id = BlockID::try_from(&[1u8; 32][..]).unwrap();
+        let index = 1;
+        let cumulative_txo_count = 1;
+        let root_element = TxOutMembershipElement::default();
+        let contents_hash = BlockContentsHash::default();
+        let id = compute_block_id(
+            version,
+            &parent_id,
+            index,
+            cumulative_txo_count,
+            &root_element,
+            &contents_hash,
+            parent_timestamp,
+        );
+        let parent_block = Block {
+            id,
+            version,
+            parent_id,
+            index,
+            cumulative_txo_count,
+            root_element,
+            contents_hash,
+            timestamp: parent_timestamp,
+        };
+        assert_matches!(
+            SgxConsensusEnclave::validate_timestamp(timestamp, &parent_block),
+            Err(Error::FormBlock(_))
+        );
+    }
+
+    #[parameterized(
+        one_ms_later_than_parent = {9877, 9876},
+        much_later_than_parent = {8765, 4567},
+        one_more_than_zero = {1, 0},
+    )]
+    fn timestamp_is_valid(timestamp: u64, parent_timestamp: u64) {
+        let version = 4;
+        let parent_id = BlockID::try_from(&[1u8; 32][..]).unwrap();
+        let index = 1;
+        let cumulative_txo_count = 1;
+        let root_element = TxOutMembershipElement::default();
+        let contents_hash = BlockContentsHash::default();
+        let id = compute_block_id(
+            version,
+            &parent_id,
+            index,
+            cumulative_txo_count,
+            &root_element,
+            &contents_hash,
+            parent_timestamp,
+        );
+        let parent_block = Block {
+            id,
+            version,
+            parent_id,
+            index,
+            cumulative_txo_count,
+            root_element,
+            contents_hash,
+            timestamp: parent_timestamp,
+        };
+        assert_eq!(
+            SgxConsensusEnclave::validate_timestamp(timestamp, &parent_block),
+            Ok(timestamp)
+        );
+    }
+
+    #[test_with_logger]
+    fn test_form_block_fails_for_invalid_timestamp(logger: Logger) {
+        let mut rng = Hc128Rng::from_seed([77u8; 32]);
+
+        for block_version in BlockVersion::iterator() {
+            let enclave = SgxConsensusEnclave::new(logger.clone());
+            let blockchain_config = BlockchainConfig {
```

Comment:
I'm not sure when we're worried about consensus working on previous block versions.
Even though previous versions wouldn't read the blocks timestamp this would still populate it for previous versions.

Reply:
Should be addressed in https://github.com/mobilecoinfoundation/mobilecoin/pull/3733/commits/5a107ab9d09b3a93e4354940ea8e61b998adbd29